### PR TITLE
Async Password Handling

### PR DESCRIPTION
**Pull request by JasonLG1979**

### CODE REVIEW

File: pithos/util.py [hunk 1]
```diff
@@ -24,76 +24,94 @@
...
+            )
+
+            if not num_items or default_collection not in unlocked:
+                self._current_collection = Secret.COLLECTION_SESSION
+                logging.debug('The default keyring is locked. Using session collection.')
+            else:
+                logging.debug('The default keyring was unlocked.')
+
+
+    def get_account_password(self, email, callback):
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E303] too many blank lines (2)

File: pithos/util.py [hunk 2]
```diff
@@ -24,76 +24,94 @@
...
+            {'email': email},
+            self._current_collection,
+            'Pandora Account',
+            password,
+            None,
+            on_password_store_finish,
+            None,
+        )
 
+SecretService = _SecretService()
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E305] expected 2 blank lines after class or function definition, found 1

File: pithos/util.py [hunk 3]
```diff
@@ -24,76 +24,93 @@
...
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E305] expected 2 blank lines after class or function definition, found 1

**Comment by JasonLG1979** [comment]
There you go @TingPing Async password handling. The keyring unlock is still sync but I want it to block so it completes before the Pithos window is shown. Anyway this also solves the problem of Pithos quitting if you change the creds and the Pandora One checkbox.

**Comment by JasonLG1979** [comment]
Amended to catch a couple edge cases.

**Comment by JasonLG1979** [comment]
@TingPing amended to be 100% async including the unlocking of the keyring.

File: pithos/util.py [hunk 4]
```diff
@@ -24,75 +24,128 @@
...
-    if previous_email and previous_email != email:
-        if not _clear_account_password(previous_email):
-            logging.warning('Failed to clear previous account')
-
-    if not password:
-        return _clear_account_password(email)
+    def set_account_password(self, email, password, callback):
+        def on_password_store_finish(source, result, data):
+            try:
+                success = Secret.password_store_finish(result)
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [F841] local variable 'success' is assigned to but never used

**Comment by JasonLG1979** [comment]
@TingPing merge this so I can work on async plugin loading next.

**Comment by TingPing** [comment]
Surely those two branches wouldn't be interdependent?

**Comment by JasonLG1979** [comment]
No but I want to see how they interact. That will make init'ing basically all async.

**Comment by JasonLG1979** [comment]
I also noticed something. We're currently trying to migrate the settings *before* we try to get the user to unlock the keyring. I'll see what I can do to fix that and amend my PR.

**Comment by JasonLG1979** [comment]
I also have another question. We have a a few plugins that use DBus and a part of their "prepare" is getting the session bus. Is there a reason we can't get the bus just once and share it among the plugins?

**Comment by JasonLG1979** [comment]
There that last little bit fixes trying to migrate setting before we try to get the user to unlock the keyring.

**Comment by TingPing** [comment]
> Is there a reason we can't get the bus just once and share it among the plugins?

I planned on sharing them at some point.

**Comment by JasonLG1979** [comment]
> I planned on sharing them at some point.

When I do the async plugin stuff I can do that. I was just wondering if there was any technical reason why we couldn't.

**Comment by JasonLG1979** [comment]
@TingPing merge this and the thumbprint radio PR so I can work on async plugins next.

File: pithos/PreferencesPithosDialog.py [hunk 5]
```diff
@@ -177,17 +177,48 @@ def on_listbox_update_header(self, row, before, junk=None):
...
+        SecretService.get_account_password(self.last_email, cb)
 
     def do_response(self, response_id):
-        if response_id == Gtk.ResponseType.APPLY:
-            self.settings.apply()
-            if set_account_password(self.settings['email'], self.password_entry.get_text(),
-                                    self.last_email):
-                self.emit('login-changed')
-        else:
+        if response_id == Gtk.ResponseType.CANCEL:
```
Comment (TingPing) [inline pithos/PreferencesPithosDialog.py]: I believe this should be left in an else because closing a dialog doesn't result in CANCEL.

Comment (JasonLG1979) [inline pithos/PreferencesPithosDialog.py]: You can't close the prefs dialog there are no window controls. The only options you have are apply or cancel

Comment (JasonLG1979) [inline pithos/PreferencesPithosDialog.py]: But it really doesn't matter. Else works just the same if that's the way you want it?

Comment (TingPing) [inline pithos/PreferencesPithosDialog.py]: You can always close a window: `alt+f4`

**Comment by JasonLG1979** [comment]
> You can always close a window: alt+f4

Point taken

File: pithos/PreferencesPithosDialog.py (hunk 5 above)
Comment (JasonLG1979) [inline pithos/PreferencesPithosDialog.py]: Point Taken

Comment (JasonLG1979) [inline pithos/PreferencesPithosDialog.py]: Done, it's an else.

**Comment by TingPing** [comment]
If you change your email it doesn't delete the old one so you have a duplicate in the keyring.

**Comment by JasonLG1979** [comment]
I will look into that. But that's kinda good that means we can support multiple accounts. Instead of having to re-enter everything we can have a email drop down that lets the user pick the account by email and have the settings auto complete by account.

**Comment by TingPing** [comment]
If you want to expose a UI for that cool, but for now we really only support one and its just leaving junk behind.

**Comment by JasonLG1979** [comment]
Right on. I can make sure it's cleared

File: pithos/util.py [hunk 6]
```diff
@@ -24,75 +24,155 @@
...
+                {'email': new_email},
+                self._current_collection,
+                'Pandora Account',
+                password,
+                None,
+                on_password_store_finish,
+                None,
+            )
+
+SecretService = _SecretService()
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E305] expected 2 blank lines after class or function definition, found 1

**Comment by JasonLG1979** [comment]
That should do it.

**Comment by TingPing** [comment]
I still get left over keys in the keyring when changing email. For context I am changing to a non-existing account so maybe you didn't handle that on failure?

**Comment by JasonLG1979** [comment]
> I still get left over keys in the keyring when changing email. For context I am changing to a non-existing account so maybe you didn't handle that on failure?

Did you clear the keyring out 1st before you tested it out?

**Comment by JasonLG1979** [comment]
What does the debug log say? I made that part very "loggy" for that reason,lol!!!

**Comment by JasonLG1979** [comment]
> so maybe you didn't handle that on failure?

It should handle failure.

```python
        def on_password_clear_finish(source, result, data):
            try:
                password_removed = Secret.password_clear_finish(result)
                if password_removed:
                    logging.debug('Cleared password for: {}'.format(old_email))
                else:
                    logging.debug('No password found to clear for: {}'.format(old_email))
            except GLib.Error as e:
                logging.error('Failed to clear previous password for: {}, Error: {}'.format(old_email, e))
                if callback:
                    callback(False)
```
This bit should just about cover anything.

**Comment by JasonLG1979** [comment]
No matter what I do, in test mode with made up creds or with real accounts I only ever have 1 entry in the keyring.

**Comment by JasonLG1979** [comment]
```
DEBUG - util:on_password_clear_finish:138 - Cleared password for: jasonlevigrayrertertert@gmail.com
```

**Comment by JasonLG1979** [comment]
Just a little wording change in on_password_clear_finish...

**Pull request merged**